### PR TITLE
add unsafe-finder tool

### PR DESCRIPTION
**Pull request by patricklam**
This pull request adds a tool to identify functions that contain unsafe code but are not unsafe, as well as pub unsafe functions.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by tautschnig** [comment]
Can you explain why this repository is the right place for such a tool? (It seems like a general Rust tool, not specific to the standard library.) We also have the `scanner` tool in the Kani repository, which produces CSV files that have (what I believe to be) equivalent information. See https://github.com/model-checking/verify-rust-std/blob/main/scripts/kani-std-analysis/std-analysis.sh or https://github.com/model-checking/kani/tree/main/tests/script-based-pre/tool-scanner.

**Comment by patricklam** [comment]
Thanks. The kani repo is perhaps a better place than the stdlib for this tool, yes.

I couldn't quite figure out how to run the scanner tool, though I have seen the kani metrics being committed to the repo. I tried to run, for instance, tests/script-based-pre/tool-scanner/scanner.test.sh and it says "cannot stat test.rs'. Running the scanner tool on, for instance, rc.rs gives a bunch of compiler errors.

Unsafe distance of 0 and 1 is similar to what my tool reports, but I only see counts as the output from the scanner tool (without running it). If it doesn't do it already, the scanner tool could also be modified to report names, like my tool does, and that might be better. I used my tool to collect the list of functions for my proposed challenges, because I figured that doing it manually was a losing game.

**Comment by patricklam** [comment]
Thanks. Yes, the std-analysis.sh script returns a superset of the information that my tool provides, but I'd suggest that it would have required some postprocessing to use it to generate the list of functions for a challenge. 

If there is interest, I could write a tool that uses the std-analysis.sh tool to produce nicely-formatted output for a challenge; it is probably better to have a single source of truth, rather than a bunch of tools which have slightly different opinions on what to report.

There's also the question of what repo this should live in. As is, the script to run the tool lives in the verify-rust-std repo, while the tool itself (which is not actually Kani-specific) lives in the kani repo. Neither of these are really ideal.

**Comment by carolynzech** [comment]
> If there is interest, I could write a tool that uses the std-analysis.sh tool to produce nicely-formatted output for a challenge; it is probably better to have a single source of truth, rather than a bunch of tools which have slightly different opinions on what to report.

Agreed. Your idea to tailor the output more closely to directories/files to make it easier to guide challenge development is a good one; the scanner tool outputs quite a bit of data right now. We'd prefer to coalesce around a single "std scanner" tool just for ease of maintenance, but we're very happy to have contributions that build on that tool. It seems like in this case, we can add some filtering to the scanner tool to reduce the output to what you're interested in.

> There's also the question of what repo this should live in. As is, the script to run the tool lives in the verify-rust-std repo, while the tool itself (which is not actually Kani-specific) lives in the kani repo. Neither of these are really ideal.

Agreed. `std-analysis.sh` used to live in Kani too; we moved it here to make it a bit less awkward. The original idea was that the scanner tool isn't specific to the standard library, so it made more sense for it to live in Kani and be more "general purpose" than to live here. I think at this point though, we only ever use it for the standard library, and having it be in the Kani repo makes it harder for other contributors such as yourself to find. So I'd be in favor of moving it here. If @tautschnig agrees, I can open a PR to do that.

**Comment by patricklam** [comment]
I've updated this tool so that it uses the output from std-analysis.sh. I noticed that kani_std_analysis.py creates a lot of graphs, but it does not produce a list of unsafe functions / unsafe abstractions, which I think is still useful. 

The tool now takes e.g. core_scan_functions.csv as output by std_analysis and produces a grouped list of functions. It would also be useful to combine it with the check for Kani annotations, I think. Is anyone else interested in this tool, possibly with the addition of a Kani filter? (There are, of course, also some things that are verified by VeriFast and by the SIMD checking approach, but they are currently not much code.) @carolynzech @tautschnig

**Comment by feliperodri** [comment]
@patricklam could you address all copilot comments before I do a full review of this PR?

File: tools/unsafe-finder/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    let mut rdr = csv::ReaderBuilder::new().delimiter(b';').from_reader(path_contents.as_bytes());
+
+    println!("# Unsafe usages in file {}", path.display());
+
+    let mut fns_by_modules: HashMap<Vec<String>, Vec<StructuredFnName>> = HashMap::new();
+    
+    for result in rdr.deserialize() {
+        let fn_stats: FnStats = result?;
+	if matches!(fn_stats.is_unsafe, Some(true)) || matches!(fn_stats.has_unsafe_ops, Some(true)) {
+	    let structured_fn_name = parse_fn_name(fn_stats.name, fn_stats.is_public.is_some() && fn_stats.is_public.unwrap());
```
[the remark this answers is not in the document]
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs]: Restructured this to align with README.md

File: tools/unsafe-finder/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        }
+    }
+
+    let mut parts_index = 0;
+    let item = &parts[parts_index]; parts_index += 1;
+    let tp = &parts[parts_index].as_str();
+    let type_parameters = if brackets_re.is_match(tp) {
+        let tp_commas = &brackets_re.captures(tp).unwrap();
+	parts_index += 1;
+        split_by_commas(&tp_commas[1]).into_iter().map(|x| x.to_string()).collect()
```
[the remark this answers is not in the document]
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs]: I couldn't find any documentation of the "established scanner format", but I now skip anything that starts with "impl".

File: tools/unsafe-finder/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    #[test]
+    fn commas_unmatched() {
+	let result = split_by_commas("<a,b),c");
+	assert_eq!(result, ["<a,b),c"]);
+    }
+}
+
+fn parse_fn_name(raw_name:String, is_public:bool) -> StructuredFnName {
+    let trait_impl_re = Regex::new(r"<(.+) as (.+)>").unwrap();
+    let brackets_re = Regex::new(r"<(.+)>").unwrap();
```
[the remark this answers is not in the document]
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs]: Fixed with a static LazyLock Regex.

File: tools/unsafe-finder/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+function; there must be some reason that the unsafe code in the
+function is actually OK. (See
+https://github.com/rust-lang/rust-clippy/issues/9330 for more
+discussion on this issue). To verify the Rust standard library, one
+must verify that there is no undefined behaviour triggered by the
+unsafe code, probably by verifying the stated reason that the code is
+OK.
+
+There are some related metrics which are automatically generated in the
+`verify-rust-std` project. Those metrics live in `scritps/kani-std-analysis/metrics-data-core.json`.
```
[the remark this answers is not in the document]
Comment (patricklam) [reply tools/unsafe-finder/README.md]: Fixed.

File: tools/unsafe-finder/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "unsafe-finder"
+version = "0.1.0"
+edition = "2024"
+
+[dependencies]
+prettyplease = "0.2.32"
+syn = {version = "2.0.101", features = ["full", "extra-traits", "visit"]}
```
[the remark this answers is not in the document]
Comment (patricklam) [reply tools/unsafe-finder/Cargo.toml]: Fixed.

File: scripts/kani-std-analysis/std-analysis.sh [hunk 6]
```diff
@@ -18,8 +20,8 @@
 set -eu
 
 # Test for platform
-PLATFORM=$(uname -sm)
-if [[ $PLATFORM == "Linux x86_64" ]]
+PLATFORM=$(uname -msp)
```
Comment (feliperodri) [inline scripts/kani-std-analysis/std-analysis.sh]: This change looks unrelated to the tool and is a regression. `uname` prints selected fields in a fixed canonical order regardless of flag order, so `uname -msp` = kernel + machine + **processor**. `uname -p` is unreliable on Linux: it returns `unknown` on some systems but the real CPU on others. On my box `uname -msp` = `"Linux x86_64 x86_64"`, so this check fails, `TARGET` is never set, and with `set -eu` the next `$TARGET` use aborts the run. Please revert to `uname -sm` / `"Linux x86_64"` (or split this into its own PR with rationale).

File: tools/unsafe-finder/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            type_parameters: vec![],
+            item: parts[0].to_string(),
+            is_public: is_public,
+            typ: typ,
+        }
+    }
+
+    let mut parts_index = 0;
+    let item = &parts[parts_index]; parts_index += 1;
+    let tp = &parts[parts_index].as_str();
```
Comment (feliperodri) [inline tools/unsafe-finder/src/main.rs:156]: Panic risk: if `raw_name` has no `::`, `parts.len() == 1`, the `parts.len() == 2` branch above is skipped, and this `parts[parts_index]` (index 1) panics with `index out of bounds`. Reproduced on a row named `foo` — it aborts the whole run. Please guard the length before indexing.

File: tools/unsafe-finder/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+fn split_by_double_colons(s:&str) -> Vec<String> {
+    let mut bracket_level = 0;
+    let mut current_string = String::new();
+    let mut previous_strings = vec![];
+    let mut colons = 0;
+    for c in s.chars() {
+	current_string.push(c);
+	match c {
+	    '<' => bracket_level += 1,
+	    '>' => bracket_level -= 1,
```
Comment (feliperodri) [inline tools/unsafe-finder/src/main.rs]: The `<`/`>` depth counter miscounts on `->`: the `>` of a fn-pointer/closure return type decrements `bracket_level` even though no `<` opened it. `mymod::<fn()->bar::Baz>::the_item` then splits into 4 parts instead of 3. Demangled std names do contain such types. Consider skipping the `>` when the previous char is `-` (and note `>>`/`>=` are also ambiguous).

File: tools/unsafe-finder/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    let parts:Vec<String> = split_by_double_colons(&raw_name).into_iter().rev().collect();
+
+    if parts.len() == 2 && TRAIT_IMPL.is_match(&parts[1]) {
+        let ti_captures = TRAIT_IMPL.captures(&parts[1]).unwrap();
+        return StructuredFnName {
+            trait_impl: Some((ti_captures[1].to_string(), ti_captures[2].to_string())),
+            module_path: vec![],
+            type_parameters: vec![],
+            item: parts[0].to_string(),
+            is_public: is_public,
```
Comment (feliperodri) [inline tools/unsafe-finder/src/main.rs]: clippy: redundant field names — `is_public: is_public` / `typ: typ` here and at 190/191 can be the shorthand `is_public,` / `typ,`. This is one of 7 clippy warnings (also an empty `else {}` at 233, a needless borrow, and an immediately-dereferenced reference); `cargo clippy --fix` clears them all.

File: tools/unsafe-finder/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "unsafe-finder"
+version = "0.1.0"
+edition = "2024"
```
Comment (feliperodri) [inline tools/unsafe-finder/Cargo.toml]: `edition = "2024"` requires cargo ≥ 1.85. It builds fine under the repo's pinned `nightly-2025-11-25`, but won't build on stable 1.84 — worth confirming that's acceptable for this tool, or dropping to 2021 since nothing here needs 2024-specific features.

**Review by feliperodri — CHANGES_REQUESTED** [review]
## Review: `add unsafe-finder tool`

**Verdict: Request changes.** This is a genuinely useful auxiliary tool — it surfaces `pub unsafe fn`s and safe-but-`unsafe`-containing functions from the scanner CSV, which is exactly the kind of prioritization the verification effort needs. It's a standalone crate under `tools/` and doesn't touch `library/`, so there's no risk to the verification target. Thank you for it. There are three blocking issues though, all reproduced locally (rustc/cargo from the repo's pinned `nightly-2025-11-25`), plus some robustness/idiom cleanups.

### Blocking
1. **The `scripts/kani-std-analysis/std-analysis.sh` change looks like an unrelated regression.** Switching `uname -sm` → `uname -msp` and matching the literal `"Linux x86_64 unknown"` hard-codes the assumption that `uname -p` returns `unknown`. On my machine `uname -msp` returns `"Linux x86_64 x86_64"` (because `uname -p` reports the real CPU there), so the platform check *fails*, `TARGET` is never set, and — with `set -eu` in force — the next use of `$TARGET` aborts the whole metrics run. This also seems out of scope for "add unsafe-finder tool." Please drop this hunk (keep the harmless comment fixes if you like) or split it into its own PR with justification.
2. **`parse_fn_name` panics on any function name that contains no `::`.** `parts[1]` is indexed unconditionally at `src/main.rs:171`. I reproduced it by feeding a row named `foo`: `thread 'main' panicked ... index out of bounds: the len is 1 but the index is 1`, which aborts the entire run. A defensive analysis tool shouldn't crash on one odd/short name — please bounds-check.
3. **`cargo fmt --check` fails** across `src/main.rs` (tabs mixed with spaces, import ordering, brace/spacing). Please run `cargo fmt`.

### Should fix
4. **The `<`/`>` depth counter in `split_by_double_colons` / `split_by_commas` breaks on `->`.** A function-pointer/closure return type inside a generic argument decrements `bracket_level` at the `>` of `->`, so nesting is miscounted. Reproduced: `mymod::<fn()->bar::Baz>::the_item` splits to `["mymod", "<fn()->bar", "Baz>", "the_item"]` instead of 3 parts. Demangled std names do contain fn-pointer/closure types, so this will misparse real rows. Consider special-casing `->` (and note `>>` shift / `>=` can't be distinguished either).
5. **7 clippy warnings** (redundant field names ×4, empty `else {}`, needless borrow, immediately-dereferenced reference). `cargo clippy --fix` clears them all.
6. **`parse_fn_name` has no unit tests** — it's the most complex and the only panic-prone function, yet only the two splitter helpers are covered (9 tests, all passing). Adding a couple of cases (trait-impl, generic, and the no-`::` edge case from item 2) would lock down the parsing.

### Minor / questions
- Every trait-impl function is bucketed under an empty `module_path` (`modules []`), so the module grouping is effectively a no-op for the majority of results. Intended?
- Category (1) is filtered to `is_public`, but category (2) reports both public and private — yet the README argues private unsafe functions also "should have contracts and be verified." Is the asymmetry deliberate?
- Leftover TODO in `main`: `// should we only handle files named "_scan_functions.csv"?`.
- Nothing invokes the tool from CI (`std-analysis.sh` generates the CSV but never calls `unsafe-finder`). Fine if it's meant to be run manually — just confirming that's the intent rather than a missing wiring step.
- `edition = "2024"` means the crate won't build on stable 1.84; it's fine under the repo's pinned nightly, just flagging for anyone building it standalone.

Happy to re-review once the blocking items are addressed, @patricklam.

File: scripts/kani-std-analysis/std-analysis.sh (hunk 6 above)
Comment (patricklam) [reply scripts/kani-std-analysis/std-analysis.sh]: I had encountered a bug with this last year, but uname -sm seems to return the right thing on my computer now. Reverting this change.

File: tools/unsafe-finder/src/main.rs (hunk 7 above)
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs:156]: Fixed & test case added.

File: tools/unsafe-finder/src/main.rs (hunk 8 above)
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs]: Added special-case code and a test. There are no instances of ">=" in the _scan_functions.csv files I've generated, so I hope we don't have to disambiguate ">>" and ">=".

File: tools/unsafe-finder/src/main.rs (hunk 9 above)
Comment (patricklam) [reply tools/unsafe-finder/src/main.rs]: Cleared.

File: tools/unsafe-finder/Cargo.toml (hunk 10 above)
Comment (patricklam) [reply tools/unsafe-finder/Cargo.toml]: Edition 2021 should be fine.

**Comment by patricklam** [comment]
Addressed all of the blocking and should-fix comments.

> Every trait-impl function is bucketed under an empty module_path (modules []), so the module grouping is effectively a no-op for the majority of results. Intended?

I forget exactly what I was thinking there, but now the module [] also says that it includes the trait impls.

> Category (1) is filtered to is_public, but category (2) reports both public and private — yet the README argues private unsafe functions also "should have contracts and be verified." Is the asymmetry deliberate?

There was a Copilot comment that led me astray, but now Category (1) includes both public and private also. I think that is the correct stance.

> Leftover TODO in main: // should we only handle files named "_scan_functions.csv"?.

It'll complain if you give it a file not named "*_scan_functions.csv" now.

> Nothing invokes the tool from CI (std-analysis.sh generates the CSV but never calls unsafe-finder). Fine if it's meant to be run manually — just confirming that's the intent rather than a missing wiring step.

I think it could be useful to hook this up to the CI, is that what we want?

File: scripts/kani-std-analysis/std-analysis.sh [hunk 11]
```diff
@@ -4,9 +4,10 @@
 
 # Collect some metrics related to the crates that compose the standard library.
 #
-# Files generates so far:
+# Files generated so far:
 #
 #  - ${crate}_scan_overall.csv: Summary of function metrics, such as safe vs unsafe.
```
Comment (tautschnig) [inline scripts/kani-std-analysis/std-analysis.sh]: This documents `_scan_functions.csv` as an output, but the change is comment-only — the script body isn't touched. Can you confirm the scanner invoked by this script actually emits `${crate}_scan_functions.csv`? The new tool consumes exactly that file, so if the pipeline doesn't produce it the README example (`unsafe-finder .../std_scan_functions.csv`) isn't reproducible.

**Review by feliperodri — APPROVED** [review]

**Comment by feliperodri** [comment]
Thank you, @patricklam!!
